libass/ass_fontselect.c

```diff
@@ -342,6 +342,65 @@ static void ass_font_provider_free_fontinfo(ASS_FontInfo *info)
     free(info->path);
 }
 
+/**
+ * \brief Fill a partly-populated ASS_FontInfo with GDI-compatible metadata.
+ * Opens the font file, reads in its metadata, and populates font->meta appropriately.
+ * On failure, font is unmodified.
+ *
+ * \param font FontInfo struct to fill
+ * \return whether the operation completed successfully
+ */
+static bool fill_font_info(ASS_FontInfo *font)
+{
+    int index = font->index;
+    ASS_FontProviderMetaData *meta = &font->meta;
+    ASS_FontProvider *provider = font->provider;
+    ASS_FontSelector *selector = provider->parent;
+    FT_Face face;
+    bool success = false;
+
+    if (provider->funcs.get_font_index)
+        index = provider->funcs.get_font_index(font->priv);
+
+    if (!font->path) {
+        assert(provider->funcs.get_data);
+        ASS_FontStream stream = {
+            .func = provider->funcs.get_data,
+            .priv = font->priv,
+        };
+        // This name is only used in an error message, so use
+        // our best name but don't panic if we don't have any.
+        // Prefer PostScript name because it is unique.
+        const char *name = meta->postscript_name ?
+            meta->postscript_name : meta->extended_family;
+        face = ass_face_stream(selector->library, selector->ftlibrary,
+                               name, &stream, index);
+    } else {
+        face = ass_face_open(selector->library, selector->ftlibrary,
+                             font->path, meta->postscript_name, index);
+    }
+
+    if (!face)
+        return false;
+
+    ASS_FontProviderMetaData local_meta = *meta;
+
+    if (!get_font_info(selector->ftlibrary, face, meta->extended_family,
+                       &local_meta))
+        goto fail;
+
+    free(meta->postscript_name);
+    local_meta.postscript_name = local_meta.postscript_name ? strdup(local_meta.postscript_name) : NULL;
+
+    memcpy(meta, &local_meta, sizeof(local_meta));
+
+    success = true;
+
+fail:
```

Comment:
Nit, subjective: I’d rather name this `cleanup`, because the code below this label executes on both failed and successful paths.

Reply:
Sure.